drivers/network/netio/netio.c

```diff
@@ -0,0 +1,896 @@
+{
+    PTDI_CONNECTION_INFORMATION TargetConnectionInfo, Ignored;
+    PIRP tdiIrp;
+    struct _WSK_SOCKET_INTERNAL *s = (struct _WSK_SOCKET_INTERNAL *)Socket;
+    NTSTATUS status;
+    struct NetioContext *nc;
+
+
+    /* Call ourselves. Sets status to pending. The interesting
+     * part happens later.
+     */
+    if (DummyDeviceObject == NULL)
+    {
+        DbgPrint("DummyDeviceObject is NULL, was the DriverEntry funtion called?\n");
+        Irp->IoStatus.Status = STATUS_INVALID_PARAMETER;
+        return STATUS_INVALID_PARAMETER;        /* TODO: something more meaningful */
+    }
+    DummyCallDriver(Irp);
+
+    nc = ExAllocatePoolWithTag(NonPagedPool, sizeof(*nc), 'NEIO');
+    if (nc == NULL)
+    {
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+    nc->socket = s;
+    nc->UserIrp = Irp;
+
+    tdiIrp = NULL;
+
+    if (s->LocalAddressHandle == INVALID_HANDLE_VALUE)
+    {
+        DbgPrint
+            ("LocalAddressHandle is not set, maybe you need to bind() your socket before connecting it?\n");
+        Irp->IoStatus.Status = STATUS_INVALID_PARAMETER;
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    status = TdiAssociateAddressFile(s->LocalAddressHandle, s->ConnectionFile);
+    if (!NT_SUCCESS(status))
+    {
+        Irp->IoStatus.Status = status;
+        return status;
+    }
+
+    TargetConnectionInfo = TdiConnectionInfoFromSocketAddress(RemoteAddress);
+    if (TargetConnectionInfo == NULL)
+    {
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    /* TODO: which address?? */
+    Ignored = TdiConnectionInfoFromSocketAddress(RemoteAddress);
+    if (Ignored == NULL)
+    {
+        ExFreePoolWithTag(TargetConnectionInfo, TAG_AFD_TDI_CONNECTION_INFORMATION);
+
+        Irp->IoStatus.Status = STATUS_INSUFFICIENT_RESOURCES;
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    /* Some drivers (like WinDRBD) do not set this.
+     * But ReactOS seems to require this (else BSOD on completion).
+     */
+
+    if (Irp->Tail.Overlay.Thread == NULL)
+    {
+        Irp->Tail.Overlay.Thread = PsGetCurrentThread();
+    }
+    IoMarkIrpPending(Irp);
+    SocketGet(s);
+    InsertTailList(&s->PendingUserIrps, &Irp->Tail.Overlay.ListEntry);
+
+    status = TdiConnect(&tdiIrp, s->ConnectionFile, TargetConnectionInfo, Ignored, NetioComplete, nc);
+    /* TODO: check this again..I would assume if the TDI helper function
+       returns any error, the completion function is not called.
+     */
+    if (!NT_SUCCESS(status))
+    {
+        SocketPut(s);
+        ExFreePoolWithTag(nc, 'NEIO');
+    }
+    return status;
+}
+
+static NTSTATUS WSKAPI
+WskStreamIo(_In_ PWSK_SOCKET Socket,
+            _In_ PWSK_BUF Buffer, _In_ ULONG Flags, _Inout_ PIRP Irp, enum direction Direction)
```

Comment:
```suggestion
    _In_ PWSK_BUF Buffer, _In_ ULONG Flags, _Inout_ PIRP Irp, enum direction Direction)
```
do not rely on the func-names length with indent.